Detectors/CPV/calib/src/NoiseCalibrator.cxx

```diff
@@ -0,0 +1,150 @@
+#include "CPVCalibration/NoiseCalibrator.h"
+#include "CommonUtils/MemFileHelper.h"
+#include "DetectorsCalibration/Utils.h"
+#include "CPVBase/Geometry.h"
+#include "CPVBase/CPVSimParams.h"
+#include "CCDB/CcdbApi.h"
+#include "CCDB/CCDBTimeStampUtils.h"
+
+namespace o2
+{
+namespace cpv
+{
+using NoiseTimeSlot = o2::calibration::TimeSlot<o2::cpv::NoiseCalibData>;
+// NoiseCalibData
+//_____________________________________________________________________________
+NoiseCalibData::NoiseCalibData()
+{
+  auto& cpvParams = o2::cpv::CPVSimParams::Instance();
+  mNoiseThreshold = cpvParams.mNoiseClbThreshold;
+  for (int i = 0; i < Geometry::kNCHANNELS; i++) {
+    mOccupancyMap.push_back(0);
+  }
+}
+//_____________________________________________________________________________
+void NoiseCalibData::fill(const gsl::span<const Digit> digits)
+{
+  for (auto& dig : digits) {
+    if (dig.getAmplitude() > mNoiseThreshold) {
+      mOccupancyMap[dig.getAbsId()]++;
+    }
+  }
+  mNEvents++;
+}
+//_____________________________________________________________________________
+void NoiseCalibData::merge(const NoiseCalibData* prev)
+{
+  for (int i = 0; i < Geometry::kNCHANNELS; i++) {
+    mOccupancyMap[i] += prev->mOccupancyMap[i];
+  }
+  mNEvents += prev->mNEvents;
+  LOG(info) << "Merged TimeSlot with previous one. Now we have " << mNEvents << " events.";
+}
+//_____________________________________________________________________________
+void NoiseCalibData::print()
+{
+  LOG(info) << "NoiseCalibData::mNEvents = " << mNEvents;
+}
+//_____________________________________________________________________________
+// NoiseCalibrator
+//_____________________________________________________________________________
+NoiseCalibrator::NoiseCalibrator()
+{
+  LOG(info) << "NoiseCalibrator::NoiseCalibrator() : "
+            << "Noise calibrator created!";
+  auto& cpvParams = o2::cpv::CPVSimParams::Instance();
+  mMinEvents = cpvParams.mNoiseClbMinEvents;
+  mToleratedChannelEfficiencyLow = cpvParams.mNoiseClbToleratedChannelEfficiencyLow;
+  mToleratedChannelEfficiencyHigh = cpvParams.mNoiseClbToleratedChannelEfficiencyHigh;
+  mNoiseFrequencyCriteria = cpvParams.mNoiseClbFrequencyCriteria;
+}
+//_____________________________________________________________________________
+void NoiseCalibrator::initOutput()
+{
+  LOG(info) << "NoiseCalibrator::initOutput() : output vectors cleared";
+  mCcdbInfoBadChannelMapVec.clear();
+  mBadChannelMapVec.clear();
+}
+//_____________________________________________________________________________
+void NoiseCalibrator::finalizeSlot(NoiseTimeSlot& slot)
+{
+  NoiseCalibData* calibData = slot.getContainer();
+  LOG(info) << "NoiseCalibrator::finalizeSlot() : finalizing slot "
+            << slot.getTFStart() << " <= TF <= " << slot.getTFEnd() << " with " << calibData->mNEvents << " events.";
+  o2::cpv::BadChannelMap* badMap = new o2::cpv::BadChannelMap();
+  bool* badMapBool = new bool[Geometry::kNCHANNELS];
```

Comment:
Why don't you use `boll badMapBool[Geometry::kNCHANNELS]={};` ? In any case, if you use `new`, you should also use `delete[]`.

Reply:
Not deleting array is clearly my mistake, thanks for pointing this out.